pkg/manager/runnable_group.go

```diff
@@ -0,0 +1,325 @@
+	group *sync.Map
+}
+
+func newRunnableGroup(errChan chan error) *runnableGroup {
+	r := &runnableGroup{
+		errChan: errChan,
+		ch:      make(chan *readyRunnable),
+		wg:      new(sync.WaitGroup),
+		group:   new(sync.Map),
+	}
+	r.ctx, r.cancel = context.WithCancel(context.Background())
+	return r
+}
+
+// Started returns true if the group has started.
+func (r *runnableGroup) Started() bool {
+	r.start.Lock()
+	defer r.start.Unlock()
+	return r.started
+}
+
+// StartAndWaitReady starts all the runnables previously
+// added to the group and waits for all to report ready.
+func (r *runnableGroup) StartAndWaitReady(ctx context.Context) error {
+	r.Start()
+	return r.WaitReady(ctx)
+}
+
+// Start starts the group, it can only be called once.
+func (r *runnableGroup) Start() {
+	r.startOnce.Do(func() {
+		go r.reconcile()
+		r.start.Lock()
+		r.started = true
+		r.group.Range(func(key, _ interface{}) bool {
+			r.ch <- key.(*readyRunnable)
+			return true
+		})
+		r.start.Unlock()
+	})
+}
+
+// reconcile is our main entrypoint for every runnable added
+// to this group. Its primary job is to read off the internal channel
+// and schedule runnables while tracking their state.
+func (r *runnableGroup) reconcile() {
+	for runnable := range r.ch {
+		// Handle stop.
+		// If the shutdown has been called we want to avoid
+		// adding new goroutines to the WaitGroup because Wait()
+		// panics if Add() is called after it.
+		{
+			r.stop.RLock()
+			if r.stopped {
+				// Drop any runnables if we're stopped.
+				r.errChan <- errRunnableGroupStopped
+				r.stop.RUnlock()
+				continue
+			}
+
+			// Why is this here?
+			// When StopAndWait is called, if a runnable is in the process
+			// of being added, we could end up in a situation where
+			// the WaitGroup is incremented while StopAndWait has called Wait(),
+			// which would result in a panic.
+			r.wg.Add(1)
+			r.stop.RUnlock()
+		}
+
+		// Start the runnable.
+		go func(rn *readyRunnable) {
+			go func() {
+				if rn.Check(r.ctx) {
+					r.group.Store(rn, true)
+				}
+			}()
+
+			// If we return, the runnable ended cleanly
+			// or returned an error to the channel.
+			//
+			// We should always decrement the WaitGroup and
+			// mark the runnable as ready.
+			//
+			// Think about the group as an append-only system.
+			//
+			// A runnable is marked as ready if:
+			// - The health check return true.
+			// - The runnable Start() method returned and
+			//	 it either finished cleanly (e.g. one shot operations)
+			//   or it failed to run and it returned an error which
+			//   gets propagated to the manager.
+			defer r.wg.Done()
+			defer r.group.Store(rn, true)
+
+			// Start the runnable.
+			if err := rn.Start(r.ctx); err != nil {
+				r.errChan <- err
+			}
+		}(runnable)
+	}
+}
+
+// WaitReady polls until the group is ready or until the context is cancelled.
+func (r *runnableGroup) WaitReady(ctx context.Context) error {
```

Comment:
Its conceptionally slightly weird that we internally have this notion of a `ready` runnable but don't make that part of our api in any way. Is it still correct that we only need this for the cache? If so, can we document it like that?

I also have to say I am not a huge fan of the poll-based architecture we end up with here and would prefer it it if we found a way to push this instead. Maybe make `reconcile` first loop over an `initial` channel and then an `additional` one for runnables that are added after the manaer is started and close a `ready` chan in between?